### PR TITLE
feat: Create table_schema api call and implementation

### PR DESCRIPTION
**Pull request by alamb**
## Rationale
In order to plan queries, DataFusion (and really any query engine) needs to know what the schemas of the input tables are

## Changes
1. adds functions to the `PartitionChunk` trait for if a table is present and what its schema is for a particular chunk / set of chunks
2. Implements this function for mutable buffer and read buffer
3. Tests that the same schema is reported for the same data laid out in several different ways

This is part of my planned sequence of PRs:
- [x] Better column selection in mutable buffer (https://github.com/influxdata/influxdb_iox/pull/700)
- [x] Support table_schema / ColumnSelection in PartitionChunk (this PR)
- [ ] Add scan_data to PartitionChunk
- [ ] Add TableProvider API (implemented in terms of scan_data and table_schema), port SQL to use

Describe your proposed changes here.

- [ ] I've read the contributing section of the project [README](https://github.com/influxdata/influxdb_iox/blob/main/README.md).
- [ ] Signed [CLA](https://influxdata.com/community/cla/) (if not already signed).

### CODE REVIEW

File: mutable_buffer/src/table.rs [hunk 1]
```diff
@@ -867,11 +869,57 @@ impl Table {
...
+        match selection {
             ChunkSelection::All => self.all_columns_selection(chunk),
             ChunkSelection::Some(cols) => self.specific_columns_selection(chunk, cols),
-        }?;
-        self.to_arrow_impl(chunk, &selection)
+        }
+    }
+
+    /// Returns the Schema of this table
+    fn schema_impl(&self, selection: &TableColSelection<'_>) -> Result<Schema> {
```
Comment (alamb) [inline mutable_buffer/src/table.rs:892]: The code here simply factors out the schema creation that was already in `to_arrow_impl` into its own function

File: read_buffer/src/lib.rs [hunk 2]
```diff
@@ -181,6 +188,66 @@ impl Database {
...
+                .iter()
+                .try_fold(SchemaMerger::new(), |builder, chunk_id| {
+                    let chunk = partition
+                        .chunks
+                        .get(chunk_id)
+                        .context(ChunkNotFound { id: *chunk_id })?;
+
+                    let result_schema = chunk.table_schema(table_name, &select_columns);
+
+                    let builder = if let Some(result_schema) = result_schema {
```
Comment (alamb) [inline read_buffer/src/lib.rs:216]: Here is the use of the schema merger to merge schemas across chunks

File: read_buffer/src/table.rs [hunk 3]
```diff
@@ -123,21 +123,27 @@ impl Table {
         // and merge results.
         let rgs = self.filter_row_groups(predicate);
 
-        let schema = ResultSchema {
+        let schema = self.result_schema(columns);
```
Comment (alamb) [inline read_buffer/src/table.rs:126]: This just factors out the creation of `ResultSchema` into a function that can be called separately

File: server/src/db.rs [hunk 4]
```diff
@@ -673,21 +671,158 @@ mod test_influxrpc {
...
+                        chunks_with_table > 0,
+                        "Expected at least one chunk to have data, but none did"
+                    );
+                }
+            }
+        };
+    }
+
+    #[tokio::test]
+    async fn list_schema_cpu_all() {
```
Comment (alamb) [inline server/src/db.rs:748]: I am quite pleased I can write tests like this -- basically test that mutable buffer and read buffer produce consistent answers

File: read_buffer/src/lib.rs [hunk 5]
```diff
@@ -181,6 +188,77 @@ impl Database {
             .sum()
     }
 
+    /// Returns the schema for the specified columns in the provided table, for
+    /// the specified partition key and chunks within that partition.
+    pub fn table_schema<'a>(
```
Comment (alamb) [inline read_buffer/src/lib.rs:193]: Here is my proposed implementation in the ReadBuffer @e-dard

File: server/src/db.rs [hunk 6]
```diff
@@ -673,21 +671,158 @@ mod test_influxrpc {
...
+        // Pick an order that is not lexographic
+        let selection = Selection::Some(&["time", "bytes"]);
+
+        run_table_schema_test_case!(TwoMeasurements {}, selection, "disk", expected_schema);
+    }
+}
+
+#[cfg(test)]
+/// This module contains testing scenarios for Db
+mod test_scenarios {
```
Comment (alamb) [inline server/src/db.rs:805]: You can tell by creating multiple modules for these sets of tests I am looking forward to some time when the tests can be broken into different files too

File: data_types/src/schema.rs [hunk 7]
```diff
@@ -394,6 +394,35 @@ impl Schema {
             Err(schema_arc) => schema_arc.as_ref().clone(),
         }
     }
+
+    /// Resort order of our columns lexographically by name
+    pub fn sort_fields_by_name(self) -> Self {
+        // paris of (orig_index, field_ref)
```
Comment (e-dard) [inline data_types/src/schema.rs:400]:
```suggestion
        // pairs of (orig_index, field_ref)
```

File: data_types/src/schema.rs [hunk 8]
```diff
@@ -394,6 +394,35 @@ impl Schema {
...
+
+        let is_sorted = sorted_fields
+            .iter()
+            .enumerate()
+            .all(|(index, pair)| index == pair.0);
+
+        if is_sorted {
+            self
+        } else {
+            // No way at present to destructure an existing Schema so
```
Comment (e-dard) [inline data_types/src/schema.rs:413]: This won't get called very often anyway will it?

File: read_buffer/src/lib.rs [hunk 9]
```diff
@@ -181,6 +188,77 @@ impl Database {
...
+                let chunk_has_table = partition
+                    .chunks
+                    .get(chunk_id)
+                    .map(|chunk| chunk.has_table(table_name))
+                    .unwrap_or(false);
+
+                if chunk_has_table {
+                    return true;
+                }
+            }
```
Comment (e-dard) [inline read_buffer/src/lib.rs:257]: I think this slightly shorter alternative does the same:

```suggestion
            return chunk_ids
                .iter()
                .any(|chunk_id| match partition.chunks.get(chunk_id) {
                    Some(chunk) => chunk.has_table(table_name),
                    None => false,
                });
```

**Review by e-dard — COMMENTED** [review]
I left a couple of suggestions. Could do with chatting about this as I want to understand the additions to the read buffer a bit more.

**Comment by e-dard** [comment]
@alamb I think I might have an alternative approach to this in terms of the read buffer.

File: data_types/src/schema.rs (hunk 8 above)
Comment (alamb) [reply data_types/src/schema.rs:413]: it will get called at least once every call to the ReadBuffer that spans multiple chunks

**Comment by alamb** [comment]
Update here: @e-dard  and I chatted a bit face to face. We discussed avoiding a potential 'lock gap' between `table_shcema` and `read_filter` calls by making a single call to get a `ReadFilterResults` (as in https://github.com/influxdata/influxdb_iox/pull/710) and use the schema from that. 

Initially, I am sure I can make that approach work, but when I thought about it some more, I think there are reasons to keep the split between "planning" and execution. 

For example, to use the DataFusion predicate and selection pushdown logic, we need the schema of the table for the plan, but we won't necessarily know what filters are actually applied during the scan until after datafusion planning is complete (after the schema was needed).

In any event I think the right thing to do is to merge #710  as soon as possible, and then I'll rebase this PR on top of that.

File: data_types/src/schema.rs (hunk 8 above)
Comment (e-dard) [reply data_types/src/schema.rs:413]: If it's limited by the number of chunks being spanned that seems fine I guess.

**Review by e-dard — APPROVED** [review]
LGTM! I already looked through this one so only needed to re-review the `ReadBuffer` bits 👍

**Comment by alamb** [comment]
FYI @tustvold

**Pull request merged**